spec/stateful/interactive.py

```diff
@@ -55,6 +55,35 @@ def syntax_error(sh):
   sh.sendline('echo status=$?')
   sh.expect('status=0')
 
+@register()
+def bg_pipeline_notify(sh):
+  'notification about background pipeline (issue 1093)'
+
+  expect_prompt(sh)
+
+  sh.sendline('sleep 0.1 | cat &')
+
+  if sh.shell_label == 'bash':
+    # e.g. [1] 12345
+    # not using trailing + because pexpect doc warns about that case
+    # dash doesn't print this
+    sh.expect(r'\[\d+\]')
+
+  expect_prompt(sh)
+
+  time.sleep(0.2)
+  sh.sendline('')
+  if 'osh' in sh.shell_label:
+      # need to wake up from wait() twice in osh
```

Comment:
Oh is this because there are 2 processes in the pipeline?

if we put 5 of them do we have to check 5 times?

Reply:
Yeah :/ We can coalesce these if we handle `SIGCHLD`. This will require some improvements to the signal handling stuff in `pyos`, through. `SIGCHLD` caries structured data that I'm sure will be handled differently between python's stdlib and libc

Reply:
Hmmm maybe we should chat about this, I'd like to understand what needs to be done -- is it a big deal?


Actually this is why I added that CheckForNotifications() method -- because I realized it wasn't enough to just  check for one

Reply:
I don't think it's a huge deal for now. And the fix might actually not even require and signal stuff. I didn't realize `main_loop::Interactive()` unconditionally waits for only one process when it gets an empty line. We could change that bit to "drain the queue". I'll push a quick patch to do that and we'll see if anything breaks...